### PR TITLE
uint256 to arith_uint256 migration (Second step).

### PR DESCRIPTION
**Pull request by furszy**
This is part of the large effort of making the uint256 file an opaque blob object and not treat every uint256 as an arith_uint256 anymore (same for the uint160 and uint512). Work coded on top of #1395, this is the second step.

As we cannot do it all at once, i opted to do the migration in steps. 
The amount of changes needed is really extensive and the risk of damage the synchronization process -and other areas as well- is high, this will need changes on the MNs, budget and zerocoin library sources which are using the uint256 object purely as a number.

This step is doing the following changes:

1) Creating blob_uint256 files (opaque blob version of the uint256 files -- this is the future uint256 files that will be refactored once all of the fields that are using it as a number in our sources get migrated to use the new arith_uint256 classes).

2) Creating arith_uint256 files.

3) Cleaning shared code between the uint256 file and the arith_uint256 classes and only leaving in uint256 the old classes that needs to be checked and migrated -- if them are using the uint256 as a number and not as an opaque blob object -- (following the explanation in point 1).

4) Updating, cleaning and connecting the arith_uint256 unit test (including arith_uint160 and arith_uint256 checks and validations).

### CODE REVIEW

**Comment by random-zebra** [comment]
Needs rebase, now that #1395 has been merged.

**Comment by furszy** [comment]
PR rebased, #1395 was merged. Ready for review.

File: src/blob_uint256.h [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+class base_blob
+{
+public:
+    // todo: make this protected
+    enum { WIDTH=BITS/8 };
+    uint8_t data[WIDTH];
+
+    base_blob()
+    {
+        memset(data, 0, sizeof(data));
```
Comment (random-zebra) [inline src/blob_uint256.h]: can't we just call `SetNull()`?

File: src/blob_uint256.h [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+    unsigned char* begin()
+    {
+        return &data[0];
+    }
+
+    unsigned char* end()
+    {
+        return &data[WIDTH];
+    }
```
Comment (random-zebra) [inline src/blob_uint256.h]: why overloading this with pointers to non-cost variables return values?
Even then, they could be made const functions (do not modify any class variable).

File: src/blob_uint256.h [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+/* uint256 from std::string.
+ * This is a separate function because the constructor uint256(const std::string &str) can result
+ * in dangerously catching uint256(0) via std::string(const char*).
+ */
+inline blob_uint256 blob_uint256S(const std::string& str)
+{
+    blob_uint256 rv;
+    rv.SetHex(str);
+    return rv;
+}
```
Comment (random-zebra) [inline src/blob_uint256.h]: we can simply `return blob_uint256S(str.c_str())`.

File: src/blob_uint256.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+template <unsigned int BITS>
+void base_blob<BITS>::SetHex(const std::string& str)
+{
+    SetHex(str.c_str());
+}
+
+template <unsigned int BITS>
+std::string base_blob<BITS>::ToString() const
+{
+    return (GetHex());
```
Comment (random-zebra) [inline src/blob_uint256.cpp]: nit: no need for the extra `()`.

File: src/blob_uint256.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    c ^= b;
+    c -= ((b << 24) | (b >> 8));
+}
+
+uint64_t blob_uint256::GetHash(const blob_uint256& salt) const
+{
+    uint32_t a, b, c;
+    const uint32_t *pn = (const uint32_t*)data;
+    const uint32_t *salt_pn = (const uint32_t*)salt.data;
+    a = b = c = 0xdeadbeef + (WIDTH << 2);
```
Comment (random-zebra) [inline src/blob_uint256.cpp]: Not sure if this should be `a = b = c = 0xdeadbeef + WIDTH;` now that WIDTH is already equal to `size()`.

File: src/blob_uint256.h (hunk 2 above)
Comment (furszy) [reply src/blob_uint256.h:64]: we have both options coded, depends on the context that its being called.

Saying that, you are pointing out something that was not changed in this PR. This two methods were moved from uint256 to blob_uint256, no behavior change.

File: src/blob_uint256.h (hunk 1 above)
Comment (furszy) [reply src/blob_uint256.h:29]: done

**Comment by furszy** [comment]
Updated.

File: src/blob_uint256.h (hunk 2 above)
Comment (random-zebra) [reply src/blob_uint256.h:64]: Ok no problem.
I just didn't see their utility, so was asking.
I know they were just moved but could have been a good occasion to change them if needed.

File: src/blob_uint256.cpp (hunk 5 above)
Comment (random-zebra) [reply src/blob_uint256.cpp:131]: How was this resolved? is it ok 128 or should it be just 32?

Comment (furszy) [reply src/blob_uint256.cpp]: now it's solved :). great catch.

Was going to appear sooner or later, unit tests are failing in debug mode. A later commit was fixing it in upstream@[5884](https://github.com/bitcoin/bitcoin/pull/5884) .

Comment (random-zebra) [reply src/blob_uint256.cpp]: Great :man_dancing:

**Comment by random-zebra** [comment]
Need to update CMakeLists with `blob_uint256` and `arith_uint256` files.

**Comment by furszy** [comment]
Done, added missing files to cmakelist.

**Pull request merged**